File: app/src/main/java/ro/alexmamo/weatherapp/city/CurrentWeatherViewModelFactory.java
package ro.alexmamo.weatherapp.city;

import android.app.Application;
import android.arch.lifecycle.ViewModel;
import android.arch.lifecycle.ViewModelProvider;
import android.support.annotation.NonNull;

import ro.alexmamo.weatherapp.cities.models.City;

@SuppressWarnings("unchecked")
public class CurrentWeatherViewModelFactory implements ViewModelProvider.Factory {
    private Application application;
    private City city;
    private CurrentWeatherRepository repository;

    CurrentWeatherViewModelFactory(
            Application application,
            City city,
            CurrentWeatherRepository repository) {
        this.application = application;
        this.city = city;
        this.repository = repository;
    }

    @NonNull
    @Override
    public <T extends ViewModel> T create(@NonNull Class<T> modelClass) {
        return (T) new CurrentWeatherViewModel(application, city, repository);
    }
}

File: app/src/main/java/ro/alexmamo/weatherapp/city/CurrentWeatherViewModelFactory.kt
package ro.alexmamo.weatherapp.city

import android.app.Application
import android.arch.lifecycle.ViewModel
import android.arch.lifecycle.ViewModelProvider

import ro.alexmamo.weatherapp.cities.models.City

@Suppress("UNCHECKED_CAST")
class CurrentWeatherViewModelFactory internal constructor(
        private val application: Application,
        private val city: City
    ) : ViewModelProvider.Factory {

    override fun <T : ViewModel> create(modelClass: Class<T>): T {
        return CurrentWeatherViewModel(application, city) as T
    }
}